rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,227 @@
+page.  This restricts what we can do with attachments.  For example,
+we cannot generate comprehensive [schema.org][] metadata for
+attachments: [see this comment][].
+
+Additionally, users of the content API cannot make use of attachments
+without parsing the page body.  This inhibits creative use of our
+content.
+
+
+## Proposal
+
+In our content schemas we already have an [`asset_link` type][], which
+is used for attachments in specialist documents and manual sections.
+It's also used in travel advice pages for the page image and a single
+attachment.  This type solves all of our problems: [Publishing API can
+render such attachments with govspeak][], and content items can expose
+this data to enable programmatic use.
+
+However, the metadata Govspeak expects and the `asset_link` type are
+*informally* the same.  And furthermore, Whitehall has much richer
+attachment metadata than either `asset_link` or Govspeak allow.
+
+Given that `asset_link` is currently used for both images and
+attachments, adding all of the missing metadata to `asset_link` feels
+a mistake: images and attachments have different metadata, so we
+shouldn't conflate the two.
+
+I propose we do this:
+
+1. Split `asset_link` into two new types: `attachment_asset` and
+   `image_asset`, updating the schemas which use `asset_link`
+   accordingly.
+
+   **Migration concerns:** This can be done without breaking
+   compatibility.
+
+1. Add new optional fields for extra metadata:
+
+   - Attachments:
+     - `accessible`
+     - `alternative_format_contact_email`
+     - `command_paper_number`
+     - `file_size`
+     - `hoc_paper_number`
+     - `isbn`
+     - `locale`
+     - `number_of_pages`
+     - `parliamentary_session`
+     - `preview_url`
+     - `unique_reference`
+     - `unnumbered_command_paper`
+     - `unnumbered_hoc_paper`
+   - Images:
+     - `alt_text`
+     - `caption`
+     - `credit`
+     - `id`
+
+   I've left out the following fields because they can be deduced from
+   others: `csv?`, `external?`, `external_url`, `file_extension`, `html?`,
+   `opendocument?`, `pdf?`
+
+   And these fields because they don't seem to be used:
+   `print_meta_data_contact_address`, `web_isbn`.
+
+   And these fields because the Publishing Workflow team are planning
+   to remove them from Whitehall: `order_url`, `price_in_pence`.
+
+   And these fields because they don't need to be exposed outside the
+   publishing app: `attachable_id`, `attachable_type`,
+   `attachment_data_id`, `deleted`, `ordering`, `slug`.
+
+   **Migration concerns:** This can be done without breaking
+   compatibility.
+
+1. Replace `content_id` in attachments with `id`.
+
+    **Migration concerns:** this would need to be done in five steps:
+    add `id`; update app to read from both fields but write to `id`;
+    republish docs; remove `content_id`; update apps to not use
+    `content_id`.
+
+1. Remove non-useful metadata fields:
+
+   - Attachments: `created_at` and `updated_at`
```

Comment:
I'm ambivalent about this change.  They don't seem to be used consistently across publishing apps: specialist-publisher sets them, travel-advice-publisher doesn't.  We could keep them though.

Reply:
I think these risk producing unhelpful or incorrect data from a consumers perspective. These fields tell us when asset was created on asset manager or updated there which is not the same as when a publisher created the asset or when they are willing to acknowledge making an edit to it. 

For example, consider migrating content from Whitehall to Content Publisher. As part of this we need to re-create the assets. This would set the metadata of an attachment as showing the asset as created at and updated at for the time of the migration. For a third party using this metadata it creates the bizarre situation that an asset published maybe 5 years ago was created and updated in 2020.

Reply:
+1 here! From my experience on DGU and GOV.UK: avoid exposing datetimes wherever possible 😬

Reply:
> This RFC looks really interesting. I had a conversation last year with the National Archives (representing other web archiving organisations) who said it would be useful to have JSON for attachments in the API responses in content items. It would allow them to collect attachments from pages.
> 
> I'll send the notes of the conversation to Michael, who can circulate it. (Apologies if this has already been mentioned in the thread above.)

Same for Parliament people who want command papers etc. 👍